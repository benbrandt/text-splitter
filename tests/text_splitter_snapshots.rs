use std::fs;

use more_asserts::assert_le;
use once_cell::sync::Lazy;
use text_splitter::{Characters, ChunkSizer, TextSplitter};
use tiktoken_rs::{cl100k_base, CoreBPE};
use tokenizers::Tokenizer;

#[test]
fn characters_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default();
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(Characters.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn characters_range() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        let splitter = TextSplitter::default();
        let chunks = splitter.chunks(&text, 500..=2000).collect::<Vec<_>>();

        assert_eq!(chunks.join(""), text);
        for chunk in chunks.iter() {
            assert_le!(Characters.chunk_size(chunk), 2000);
        }
        insta::assert_yaml_snapshot!(chunks);
    });
}

#[test]
fn characters_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default().with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static HUGGINGFACE_TOKENIZER: Lazy<Tokenizer> =
    Lazy::new(|| Tokenizer::from_pretrained("bert-base-cased", None).unwrap());

#[test]
fn huggingface_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(HUGGINGFACE_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn huggingface_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static TIKTOKEN_TOKENIZER: Lazy<CoreBPE> = Lazy::new(|| cl100k_base().unwrap());

#[test]
fn tiktoken_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(TIKTOKEN_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn tiktoken_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}
