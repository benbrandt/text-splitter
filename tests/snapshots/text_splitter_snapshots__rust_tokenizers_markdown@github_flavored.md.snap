---
source: tests/text_splitter_snapshots.rs
assertion_line: 324
expression: chunks
input_file: tests/inputs/markdown/github_flavored.md
---
- "# Headers\n\n"
- "```\n"
- "# h1 Heading 8-)\n## h2 Heading\n"
- "### h3 Heading\n#### h4 Heading\n"
- "##### h5 Heading\n"
- "###### h6 Heading\n\n"
- "Alternatively, for H1 and H2, an underline-ish style:\n\n"
- "Alt-H1\n======\n\nAlt-H2\n"
- "------\n```\n\n"
- "# h1 Heading 8-)\n"
- "## h2 Heading\n"
- "### h3 Heading\n"
- "#### h4 Heading\n"
- "##### h5 Heading\n"
- "###### h6 Heading\n\n"
- "Alternatively, for H1 and H2, an underline-ish style:\n\n"
- "Alt-H1\n======\n\n"
- "Alt-H2\n------\n\n------\n\n"
- "# Emphasis\n\n"
- "```\n"
- "Emphasis, aka italics, with *asterisks* or "
- "_underscores_.\n\n"
- "Strong emphasis, aka bold, with **asterisks** or "
- "__underscores__.\n\n"
- Combined emphasis with **asterisks and _underscores_*
- "*.\n\nStrikethrough uses two tildes. ~~Scratch this.~~\n\n"
- "**This is bold text**\n\n__This is bold text__\n\n"
- "*This is italic text*\n\n_This is italic text_\n\n"
- "~~Strikethrough~~\n```\n\n"
- "Emphasis, aka italics, with *asterisks* or "
- "_underscores_.\n\n"
- "Strong emphasis, aka bold, with **asterisks** or "
- "__underscores__.\n\n"
- "Combined emphasis with "
- "**asterisks and _underscores_**.\n\n"
- "Strikethrough uses two tildes. ~~Scratch this.~~\n"
- "\n**This is bold text**\n\n__This is bold text__\n"
- "\n*This is italic text*\n\n_This is italic text_\n"
- "\n~~Strikethrough~~\n\n------\n\n"
- "# Lists\n\n"
- "```\n"
- "1. First ordered list item\n2. Another item\n"
- "⋅⋅* Unordered sub-list.\n1. "
- "Actual numbers don't matter, just that it's a number\n"
- "⋅⋅1. Ordered sub-list\n4. And another item.\n\n"
- "⋅⋅⋅You can have properly indented paragraphs within list items. "
- "Notice the blank line above, and the leading spaces (at least one, but "
- "we'll use three here to also align the raw Markdown).\n\n"
- "⋅⋅⋅To have a line break without a paragraph, you will need to use two trailing "
- "spaces.⋅⋅\n"
- "⋅⋅⋅Note that this line is separate, but within the same paragraph.⋅⋅\n"
- "⋅⋅⋅(This is contrary to the typical GFM line break behaviour, where"
- " trailing spaces are not required.)\n\n* Unordered list can use asterisks\n"
- "- Or minuses\n+ Or pluses\n\n1. Make my changes\n    1. "
- "Fix bug\n    2. Improve formatting\n        - Make the headings bigger\n2. "
- "Push my commits to GitHub\n3. Open a pull request\n    * Describe my changes\n"
- "    * Mention all the members of my team\n        * Ask for feedback\n\n"
- "+ Create a list by starting a line with `+`, `-`"
- ", or `*`\n"
- "+ Sub-lists are made by indenting 2 spaces:\n"
- "  - Marker character change forces new list start:\n"
- "    * Ac tristique libero volutpat at\n"
- "    + Facilisis in pretium nisl aliquet\n"
- "    - Nulla volutpat aliquam velit\n+ Very easy!\n"
- "```\n\n1. First ordered list item\n"
- "2. Another item\n⋅⋅* Unordered sub-list.\n"
- "1. Actual numbers don't matter, just that it's a number\n"
- "⋅⋅1. Ordered sub-list\n4. And another item.\n\n"
- "⋅⋅⋅You can have properly indented paragraphs within list items. "
- "Notice the blank line above, and the leading spaces (at least one, but "
- "we'll use three here to also align the raw Markdown).\n\n"
- "⋅⋅⋅To have a line break without a paragraph, you will need to use two trailing "
- "spaces.⋅⋅\n"
- "⋅⋅⋅Note that this line is separate, but within the same paragraph.⋅⋅\n"
- "⋅⋅⋅(This is contrary to the typical GFM line break behaviour, where"
- " trailing spaces are not required.)\n\n* Unordered list can use asterisks\n"
- "- Or minuses\n+ Or pluses\n\n"
- 1. Make my changes
- "\n    1. Fix bug\n    2. Improve formatting\n        - Make the headings bigger\n"
- "2. Push my commits to GitHub\n"
- 3. Open a pull request
- "\n    * Describe my changes\n    * Mention all the members of my team\n        * Ask for feedback\n\n"
- "+ Create a list by starting a line with `+`, `-`"
- ", or `*`\n"
- "+ Sub-lists are made by indenting 2 spaces:"
- "\n  - Marker character change forces new list start:"
- "\n    * Ac tristique libero volutpat at\n"
- "    + Facilisis in pretium nisl aliquet\n"
- "    - Nulla volutpat aliquam velit\n+ Very easy!\n\n"
- "------\n\n"
- "# Task lists\n\n"
- "```\n"
- "- [x] Finish my changes\n- [ ] Push my commits to GitHub\n"
- "- [ ] Open a pull request\n"
- "- [x] @mentions, #refs, [links]()"
- ", **formatting**, and <del>tags</del"
- "> supported\n"
- "- [x] list syntax required (any unordered or ordered list supported"
- ")\n- [x] this is a complete item\n"
- "- [ ] this is an incomplete item\n```\n\n"
- "- [x] Finish my changes\n- [ ] Push my commits to GitHub\n"
- "- [ ] Open a pull request\n"
- "- [x] @mentions, #refs, [links]()"
- ", **formatting**, and <del>tags"
- "</del> supported\n"
- "- [x] "
- "list syntax required (any unordered or ordered list supported)\n"
- "- [ ] this is a complete item\n- [ ] this is an incomplete item\n\n"
- "------\n\n"
- "# Ignoring Markdown formatting\n\n"
- "You can tell GitHub to ignore (or escape) Markdown formatting by using \\ "
- "before the Markdown character.\n\n"
- "```\n"
- "Let's rename \\*our-new-project\\* to \\"
- "*our-old-project\\*.\n```\n\n"
- "Let's rename \\*our-new-project\\* to \\"
- "*our-old-project\\*.\n\n------\n\n"
- "# Links\n\n"
- "```\n"
- "[I'm an inline-style link](https://"
- "www.google.com)\n\n"
- "[I'm an inline-style link with title](https:/"
- "/www.google.com \"Google's Homepage\")\n\n"
- "[I'm a reference-style link][Arbitrary case-"
- "insensitive reference text]\n\n"
- "[I'm a relative reference to a repository file](../"
- "blob/master/LICENSE)\n\n"
- "[You can use numbers for reference-style link definitions][1]\n\n"
- "Or leave it empty and use the [link text itself].\n\n"
- "URLs and URLs in angle brackets will automatically get turned into links.\n"
- "http://www.example.com or <http://"
- "www.example.com> and sometimes\n"
- "example.com (but not on Github, for example).\n\n"
- "Some text to show that the reference links can follow later.\n\n"
- "[arbitrary case-insensitive reference text]: https://"
- "www.mozilla.org\n"
- "[1]: http://slashdot.org\n"
- "[link text itself]: http://www.reddit.com\n"
- "```\n\n"
- "[I'm an inline-style link](https://"
- "www.google.com)\n\n"
- "[I'm an inline-style link with title](https:/"
- "/www.google.com \"Google's Homepage\")\n\n"
- "[I'm a reference-style link][Arbitrary case-"
- "insensitive reference text]\n\n"
- "[I'm a relative reference to a repository file](../"
- "blob/master/LICENSE)\n\n"
- "[You can use numbers for reference-style link definitions][1]\n"
- "\nOr leave it empty and use the [link text itself].\n"
- "\nURLs and URLs in angle brackets will automatically get turned into links.\n"
- "http://www.example.com or "
- "<http://www.example.com> and sometimes\n"
- "example.com (but not on Github, for example).\n\n"
- "Some text to show that the reference links can follow later.\n"
- "\n"
- "[arbitrary case-insensitive reference text]: https://"
- "www.mozilla.org\n"
- "[1]: http://slashdot.org\n"
- "[link text itself]: http://www.reddit.com\n\n"
- "------\n\n"
- "# Images\n\n"
- "```\n"
- "Here's our logo (hover to see the title text):\n\n"
- "Inline-style:\n!["
- "alt text](https://github.com/adam-"
- p/markdown-here/raw/master/src/common/
- "images/icon48.png \"Logo Title Text 1\")\n\n"
- "Reference-style:\n![alt text][logo]\n\n"
- "[logo]: https://github.com/adam-"
- p/markdown-here/raw/master/src/common/
- "images/icon48.png \"Logo Title Text 2\"\n\n!["
- "Minion](https://octodex.github.com"
- "/images/minion.png)\n!["
- "Stormtroopocat](https://octodex.github.com"
- "/images/stormtroopocat.jpg \"The Stormtroopocat\")\n\n"
- "Like links, Images also have a footnote style syntax\n\n!["
- "Alt text][id]\n\n"
- "With a reference later in the document defining the URL location:\n\n"
- "[id]: https://"
- octodex.github.com/images/
- "dojocat.jpg  \"The Dojocat\"\n```\n\n"
- "Here's our logo (hover to see the title text):\n"
- "\nInline-style:\n"
- "!["
- alt text
- "](https://github.com/adam-p/"
- markdown-here/raw/master/src/common/images/
- "icon48.png \"Logo Title Text 1\")\n\n"
- "Reference-style:\n![alt text][logo]\n"
- "\n"
- "[logo]: https://github.com/adam-"
- p/markdown-here/raw/master/src/common/
- "images/icon48.png \"Logo Title Text 2\"\n\n"
- "!["
- Minion
- "](https://octodex.github.com/"
- "images/minion.png)\n"
- "!["
- Stormtroopocat
- "](https://octodex.github.com/"
- "images/stormtroopocat.jpg \"The Stormtroopocat\")\n\n"
- "Like links, Images also have a footnote style syntax\n"
- "\n![Alt text][id]\n"
- "\nWith a reference later in the document defining the URL location:\n"
- "\n"
- "[id]: https://"
- octodex.github.com/images/
- "dojocat.jpg  \"The Dojocat\"\n\n------\n\n"
- "# "
- "[Footnotes](https://github.com/markdown"
- "-it/markdown-it-footnote)\n\n"
- "```\n"
- "Footnote 1 link[^first].\n\nFootnote 2 link[^second].\n\n"
- "Inline footnote^[Text of inline footnote] definition.\n\n"
- "Duplicated footnote reference[^second].\n\n"
- "[^first]: Footnote **can have markup**\n\n"
- "    and multiple paragraphs.\n\n[^second]: Footnote text.\n```\n\n"
- "Footnote 1 link[^first].\n\nFootnote 2 link[^second].\n"
- "\nInline footnote^[Text of inline footnote] definition.\n"
- "\nDuplicated footnote reference[^second].\n"
- "\n[^first]: Footnote **can have markup**\n\n"
- "    and multiple paragraphs.\n\n[^second]: Footnote text.\n\n"
- "------\n\n"
- "# Code and Syntax Highlighting\n\n"
- "```\n"
- "Inline `code` has `back-ticks around` it.\n"
- "```\n\n"
- "Inline `code` has `back-ticks around` it.\n"
- "\n```c#\n"
- "using System.IO.Compression;\n\n"
- "#pragma warning disable 414, 3021\n\nnamespace MyApplication\n{\n"
- "    [Obsolete(\"...\")]\n    class Program : IInterface\n    {\n"
- "        public static List<int> JustDoIt(int count)\n        {\n"
- "            Console.WriteLine($\"Hello {Name}!\");\n"
- "            return new List<int>(new int[] { 1, 2, "
- "3 })\n        }\n    }\n}\n```\n\n"
- "```css\n"
- "@font-face {\n"
- "  font-family: Chunkfive; src: url('"
- "Chunkfive.otf');\n}\n\nbody, .usertext {\n"
- "  color: #F0F0F0; background: #600;\n"
- "  font-family: Chunkfive, sans;\n}\n\n"
- "@import url(print.css);\n@media print {\n"
- "  a[href^=http]::after {\n"
- "    content: attr(href)\n  }\n}\n```\n\n"
- "```javascript\n"
- "function $initHighlight(block, cls) {\n  try {\n"
- "    if (cls.search(/\\bno\\-highlight\\b"
- "/) != -1)\n"
- "      return process(block, true, 0x0F) +\n"
- "             ` class=\"${cls}\"`;\n"
- "  } catch (e) {\n    /* handle exception */\n  }\n"
- "  for (var i = 0 / 2; i < classes.length; i"
- "++) {\n"
- "    if (checkCondition(classes[i]) === undefined)\n"
- "      console.log('undefined');\n  }\n}\n\nexport  $initHighlight;\n"
- "```\n\n"
- "```php\n"
- "require_once 'Zend/Uri/Http.php';\n\n"
- "namespace Location\\Web;\n\ninterface Factory\n{\n    static function _factory();\n"
- "}\n\nabstract class URI extends BaseURI implements Factory\n{\n    abstract function test();\n\n"
- "    public static $st1 = 1;\n    const ME = \"Yo\";\n"
- "    var $list = NULL;\n    private $var;\n\n    /**\n"
- "     * Returns a URI\n     *\n     * @return URI\n     */\n"
- "    static public function _factory($stats = array(), $uri "
- "= 'http')\n    {\n        echo __METHOD__;\n"
- "        $uri = explode(':', $uri, "
- "0b10);\n"
- "        $schemeSpecific = isset($uri[1]) ? "
- "$uri[1] : '';\n        $desc = 'Multi\n"
- "line description';\n\n        // Security check\n        if (!"
- "ctype_alnum($scheme)) {\n"
- "            throw new Zend_Uri_Exception('Illegal scheme');\n        }\n\n"
- "        $this->var = 0 - self::$st;\n"
- "        $this->list = list(Array(\"1\"=> 2"
- ", 2=>self::ME, 3 => \\Location\\Web"
- "\\URI::class));\n\n        return [\n"
- "            'uri'   => $uri,\n"
- "            'value' => null,\n        ];\n    }\n}\n\n"
- "echo URI::ME . URI::$st1;\n\n"
- "__halt_compiler () ; datahere\ndatahere\n"
- "datahere */\ndatahere\n```\n\n"
- "------\n\n"
- "# Tables\n\n"
- "```\n"
- "Colons can be used to align columns.\n\n| Tables        | Are           | Cool  |\n"
- "| ------------- |:"
- "-------------:| -"
- "----:|\n"
- "| col 3 is      | right-aligned | $1600 |\n"
- "| col 2 is      | centered      |   $12 |\n"
- "| zebra stripes | are neat      |    $1 |\n\n"
- "There must be at least 3 dashes separating each header cell.\n"
- "The outer pipes (|) are optional, and you don't need to "
- "make the\nraw Markdown line up prettily. "
- "You can also use inline Markdown.\n\nMarkdown | Less | Pretty\n"
- "--- | --- | ---\n"
- "*Still* | `renders` | **nicely**\n"
- "1 | 2 | 3\n\n| First Header  | Second Header |\n"
- "| ------------- | -"
- "------------ |\n"
- "| Content Cell  | Content Cell  |\n| Content Cell  | Content Cell  |\n\n"
- "| Command | Description |\n| --- | --- |\n"
- "| git status | List all new or modified files |\n"
- "| git diff | Show file differences that haven't been staged |\n\n"
- "| Command | Description |\n| --- | --- |\n"
- "| `git status` | List all *new or modified* files |\n"
- "| `git diff` | Show file differences that **"
- "haven't been** staged |\n\n"
- "| Left-aligned | Center-aligned | Right-aligned |\n"
- "| :---         |     :---:      |          ---: "
- "|\n| git status   | git status     | git status    |\n"
- "| git diff     | git diff       | git diff      |\n\n"
- "| Name     | Character |\n| ---      | ---       |\n"
- "| Backtick | `         |\n| Pipe     | \\|        |\n```\n\n"
- "Colons can be used to align columns.\n"
- "\n| Tables        | Are           | Cool  |\n"
- "| ------------- |:"
- "-------------:| -"
- "----:|\n"
- "| col 3 is      | right-aligned | $1600 |\n"
- "| col 2 is      | centered      |   $12 |\n"
- "| zebra stripes | are neat      |    $1 |\n"
- "\nThere must be at least 3 dashes separating each header cell.\n"
- "The outer pipes (|) are optional, and you don'"
- "t need to make the\n"
- "raw Markdown line up prettily. You can also use inline Markdown.\n\n"
- "Markdown | Less | Pretty\n--- | --- | ---\n"
- "*Still* | `renders` | **nicely**\n"
- "1 | 2 | 3\n"
- "\n| First Header  | Second Header |\n"
- "| ------------- | -"
- "------------ |\n"
- "| Content Cell  | Content Cell  |\n| Content Cell  | Content Cell  |\n"
- "\n| Command | Description |\n| --- | --- |\n"
- "| git status | List all new or modified files |\n"
- "| git diff | Show file differences that haven't been staged |\n"
- "\n| Command | Description |\n| --- | --- |\n"
- "| `git status` | List all *new or modified* files |\n"
- "| `git diff` |"
- " Show file differences that **haven't been** staged |\n\n"
- "| Left-aligned | Center-aligned | Right-aligned |\n"
- "| :---         |     :---:      |          ---: "
- "|\n| git status   | git status     | git status    |\n"
- "| git diff     | git diff       | git diff      |\n"
- "\n| Name     | Character |\n| ---      | ---       |\n"
- "| Backtick | `         |\n| Pipe     | \\|        |\n\n"
- "------\n\n"
- "# Blockquotes\n\n"
- "```\n"
- "> Blockquotes are very handy in email to emulate reply text.\n"
- "> This line is part of the same quote.\n\nQuote break.\n\n"
- "> This is a very long line that will still be quoted properly when it wraps"
- ". "
- "Oh boy let's keep writing to make sure this is long enough to actually "
- "wrap for everyone. "
- "Oh, you can *put* **Markdown** into a "
- "blockquote.\n\n> Blockquotes can also be nested...\n"
- ">> ...by using additional greater-than signs right next to each "
- "other...\n> > > ...or with spaces between arrows.\n"
- "```\n\n"
- "> Blockquotes are very handy in email to emulate reply text.\n> "
- "This line is part of the same quote.\n\nQuote break.\n"
- "\n> "
- "This is a very long line that will still be quoted properly when it wraps. "
- "Oh boy let'"
- "s keep writing to make sure this is long enough to actually wrap for everyone. "
- "Oh, you can *put* **Markdown**"
- " into a blockquote.\n\n"
- "> Blockquotes can also be nested...\n>"
- "> ...by using additional greater-than signs right next to each other"
- "...\n> > > ...or with spaces between arrows.\n\n"
- "------\n\n"
- "# Inline HTML\n\n"
- "```\n"
- "<dl>\n  <dt>Definition list</dt>\n"
- "  <dd>Is something people use sometimes.</dd>\n\n"
- "  <dt>Markdown in HTML</dt>\n"
- "  <dd>Does *not* work **very** well. "
- "Use HTML <em>tags</em>.</dd>\n"
- "</dl>\n```\n\n<dl>\n"
- "  <dt>Definition list</dt>\n"
- "  <dd>Is something people use sometimes.</dd>\n"
- "\n  <dt>Markdown in HTML</dt>\n"
- "  <dd>Does *not* work **very** well. "
- "Use HTML <em>tags</em>.</dd>\n"
- "</dl>\n\n------\n\n"
- "# Horizontal Rules\n\n"
- "```\n"
- "Three or more...\n\n---\n\nHyphens\n\n***\n\nAsterisks\n\n"
- "___\n\nUnderscores\n```\n\nThree or more...\n\n---\n"
- "\nHyphens\n\n***\n\nAsterisks\n\n___\n\nUnderscores\n\n------\n\n"
- "# YouTube Videos\n\n"
- "```\n"
- "<a href=\"http://www.youtube.com/"
- watch?
- "feature=player_embedded&v=YOUTUBE_VIDEO_ID_HERE\" "
- "target=\"_blank\">\n"
- "<img src=\"http://"
- img.youtube.com/vi/YOUTUBE_VIDEO_ID_HERE
- "/0.jpg\" alt=\"IMAGE ALT TEXT HERE\" width="
- "\"240\" height=\"180\" border=\"10\">\n"
- "</a>\n```\n\n"
- "<a href=\"http://www.youtube.com/"
- watch?
- "feature=player_embedded&v=YOUTUBE_VIDEO_ID_HERE\" "
- "target=\"_blank\">\n"
- "<img src=\"http://"
- img.youtube.com/vi/YOUTUBE_VIDEO_ID_HERE
- "/0.jpg\" alt=\"IMAGE ALT TEXT HERE\" width="
- "\"240\" height=\"180\" border=\"10\">\n"
- "</a>\n"
- "\n```\n"
- "[!["
- "IMAGE ALT TEXT HERE](http://img.youtube.com"
- /vi/YOUTUBE_VIDEO_ID_HERE/0.jpg)
- "](http://www.youtube.com/watch?"
- "v=YOUTUBE_VIDEO_ID_HERE)\n```\n\n"
- "[!["
- IMAGE ALT TEXT HERE
- "](https://upload.wikimedia.org/"
- wikipedia/commons/thumb/e/ef/YouTube_logo_2015
- ".svg/1200px-YouTube_logo_2015."
- "svg.png)](https://"
- "www.youtube.com/watch?v=ciawICBvQoE)\n"
